include/oneapi/dpl/pstl/histogram_impl.h

```diff
@@ -47,8 +47,32 @@ __pattern_histogram(_Tag, _ExecutionPolicy&& exec, _RandomAccessIterator1 __firs
 } // namespace __internal
 
 template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _RandomAccessIterator2,
-          typename _ValueType>
+          typename> //final unused template param support extension API in combination with overload below
```

Comment:
I do not see why this extra template parameter is needed. As far as I know, it is fine to have function template overloads with different numbers of template parameters.

Reply:
In combination with the other overload, this unused defaulted template parameters covers the case where a user specified all template parameters explicitly, and the specified type matches the input iterator value type.

`histogram<Policy, int*, std::size_t, uint32_t*, int>(policy, first, last, num_bins, bin_min, bin_max, histogram_first);`

The other overload would exclude it because of the matching type.

Reply:
Since C++20, the standard has a clause in [algorithms.requirements] similar to this (taken from the current draft):

> The well-formedness and behavior of a call to an algorithm with an explicitly-specified template argument list is unspecified, except where explicitly stated otherwise.
[Note 3: Consequently, an implementation can declare an algorithm with different template parameters than those presented. — end note]

I believe this was added due to range algorithms, which are usually implemented in a way that they cannot be called with explicit template arguments. The restriction however, as I read it, is not limited to range algorithms.

Since we are adding parallel range algorithms both to the library and the spec as well, we will need a similar clause, and maybe it makes sense to apply to all algorithms.

Reply:
Very interesting, I wasn't aware of that.